### ISSUE TITLE
Proposal: Post-Render Args or Env Vars

### ISSUE THREAD

**Issue by acaire** [helm/helm#7849]
Post render functionality implemented in #7260 is awesome (thanks @thomastaylor312).  I (want to) manage my helm charts with Helmfile and Kustomize and don't want to rely on having to necessarily parse manifests from `stdin` or duplicate the render script for each app, instead favouring something simple and generic like so (thanks again @thomastaylor312 😆 ):
```
#!/bin/bash
set -eu
cat <&0 > $1/all.yaml
kustomize build $1 && rm $1/all.yaml
```

With an example file structure like:
```
helmfile.yaml
redis/kustomization.yaml
redis/patch.yaml
app2/...
app3/...
```

example `kustomization.yaml`:
```
resources:
- all.yaml
patches:
- patch.yaml
```

I therefore propose either being able to pass arguments or environment variables to post render execution, thus avoiding [some others kinds of hackiness](https://github.com/roboll/helmfile#helmfile--kustomize)

**Comment by thomastaylor312** [comment]
Hey @acaire thanks for the proposal! Can you document a bit more what the user interaction would look like from the CLI perspective? I think I have it figured out by looking at code, but I would like to see it documented in this proposal so people can comment on it from the UI side of things.

My initial worry here is that we are reinventing full argument parsing to do this, which feels a bit dangerous, but the feature does seem useful

**Comment by acaire** [comment]
Thanks! And agreed. I came up with Env vars as an alternative method right before submission but had already created the proof of concept.  It's probably a better fit IMHO. Regardless, I'll work on a stronger use case and proposal and update back here in a day or two.

**Comment by acaire** [comment]
@thomastaylor312, here's a practical example of what I'm trying to solve.  I'm using Helmfile which effectively wraps this command:
```
helm diff upgrade --allow-unreleased redis stable/redis
```

This shows me what the rendered chart template would look like.

Let's say for example instead of 2 replicas, I want 200 (for simplicity sake, pretend I'm a YAML purist) except I don't want to fork the chart. So I elect to use Kustomize instead (as per your examples that I covered in the first comment).

So now I'll run:
```
helm diff upgrade --allow-unreleased redis stable/redis --post-renderer kustomize.sh
```

Which is great for this app, but I store all of my kustomizations in one repo and obviously don't want to run the same redis patch for all of my other apps. So i'm forced to either:
1. Duplicate `kustomize.sh` for every instantiation of every chart, perhaps instead calling it `kustomize-redis.sh`, or even `kustomize-redis-test.sh` and `kustomize-redis-prod.sh`
or
2. Build my application logic into `kustomize.sh`

Both of these options seem terrible, because my source of truth is `helmfile.yaml`, and I have no desire to wrap the wrapper, i.e.:
```
# pseudo
for app in $(grep app helmfile.yaml):
   helmfile sync app
done
```
Because Helmfile already has a way to compile a list of all releases I can already just run this:
```
helmfile sync
```

But the current implementation of Helmfile, sees me using hooks that would download the chart, patch it, and then use it as a source, and then repeat this hackiness for every helm action (diff, upgrade etc).

Instead, if `--post-renderer` could be passed arguments (or environment variables, or some way to notify a different patch is required), the transform functionality you added into Helm could flex it's muscles, and remove existing hackiness from other projects.
So perhaps i'd split on commas:
```
helm diff upgrade --allow-unreleased redis stable/redis --post-renderer kustomize.sh,prod,redis
```
Or use a new flag and split that on commas:
```
helm diff upgrade --allow-unreleased redis stable/redis --post-renderer kustomize.sh --post-renderer-args prod,redis
```
Or expose some environment variables:
```
helm diff upgrade --allow-unreleased redis stable/redis --post-renderer kustomize.sh --post-renderer-env ENV=prod,APP=redis
```
Or maybe (but hopefully not) just pass on already exported variables :cry::
```
ENV=prod APP=redis helm diff upgrade --allow-unreleased redis stable/redis --post-renderer kustomize.sh --post-renderer kustomize.sh --post-renderer-expose-envs
```
Perhaps taking it one step further I could specify the output format (i.e. JSON passes in the command that was executed, the chart name values etc):
```
helm diff upgrade --allow-unreleased redis stable/redis --post-renderer kustomize.sh --post-renderer-output=json # defaults instead to raw
```

So depending on how this is implemented, I need to go back to helm-diff and suggest changes to the [outstanding PR](https://github.com/databus23/helm-diff/pull/196) and also go and submit a PR to [Helmfile](https://github.com/roboll/helmfile).

The power of `--post-renderer` is incredible.  Helmfile, or `<Insert wrapper here>` could go back to focusing on how best to wrap Helm for 1000s of releases in an intuitive way, and leave the heavy lifting to the renderer to reimplement things such as secret fetching and caching (instead of running x number of variable [SSM](https://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-parameter-store.html) parameter fetches for each of my 1000 releases, fetching all the parameter secrets in much fewer API calls).

In a perfect world, I wouldn't need Helmfile or Kustomize, and Helm would simply do all of the heavy lifting for me as defined via:
```
# helm.yaml
releases:
- name: redis
  chart: stable/redis
  patches:
  - target:
      apiVersion: apps/v1
      kind: StatefulSet
      metadata:
        name: redis-slave
      spec:
        replicas: 250
```

And executed via:
```
helm sync
```

Which would cover the use case of every business I've ever worked at, without having to constantly reinvent the wheel with this twisted web of shell wrapping.
So, until that happens (can we make it happen ? 😉) , I'm trying to reduce the dependency and wrapping kludge as much as possible.

**Comment by yurrriq** [comment]
+:100: We use helmfile too and have pretty much exactly the use case @acaire describe in detail.

**Comment by thomastaylor312** [comment]
@acaire Sorry for the delay on this, got really busy with some other projects. I personally like the ENV variable option the best as it is the most clean. Another option here if command line args are the most desired option, then we could use the `--` delimiter and everything else is passed as args to post render. However, I still think the variables is the best. I am going to ping the other maintainers to get a second opinion before you do any work though

**Comment by unguiculus** [comment]
@thomastaylor312 I like both ideas, env variables and passing args after `--`. Why not add both?

**Comment by thomastaylor312** [comment]
![tenor-58280278](https://user-images.githubusercontent.com/4294228/81873169-64331d00-9538-11ea-8a87-f3aa73fe6304.gif)

**Comment by paulczar** [comment]
I like both.  I really like the `--` separation to set args to the second app, but I think env vars might be friendlier to people glueing apps together in something like `helmfile`.

Would it be really crazy to process templates in a postrender dir?  like if I'm using `ytt` and I want to turn helm values into `ytt` values I could have a `ytt-values.yaml.tmpl` in my postrender dir (then pass ``-- -f ytt/ytt-values.yaml` to the post-render script) that would be rendered out and usable by the postrender script ...  or is that maybe a job we could pass upstream to helmfile and similar to perform ?

**Comment by thomastaylor312** [comment]
Yeah, I am thinking that might be above and beyond what helm should do in this case

**Comment by acaire** [comment]
Thanks for the feedback! I’ll modify the PR accordingly.

**Comment by Chili-Man** [comment]
I'm still interested in this feature; is there anyway I can help push this through

**Comment by faizanf47** [comment]
Still interested in this feature.
